crates/fuel-core/src/schema/coins.rs

```diff
@@ -233,75 +298,187 @@ impl CoinQuery {
-        });
-
-        let base_asset_id = params.base_asset_id();
-        let spend_query =
-            SpendQuery::new(owner, &query_per_asset, excluded_ids, *base_asset_id)?;
-
-        let query = ctx.read_view()?;
-
-        let coins = random_improve(query.as_ref(), &spend_query)
-            .await?
-            .into_iter()
-            .map(|coins| {
-                coins
-                    .into_iter()
-                    .map(|coin| match coin {
-                        coins::CoinType::Coin(coin) => CoinType::Coin(coin.into()),
-                        coins::CoinType::MessageCoin(coin) => {
-                            CoinType::MessageCoin(coin.into())
-                        }
-                    })
-                    .collect_vec()
-            })
-            .collect();
-
-        Ok(coins)
+        let read_view = ctx.read_view()?;
+        let indexation_available = read_view.coins_to_spend_indexation_enabled;
+        if indexation_available {
+            coins_to_spend_with_cache(
+                owner,
+                query_per_asset,
+                excluded_ids,
+                max_input,
+                read_view.as_ref(),
+            )
+            .await
+        } else {
+            let base_asset_id = params.base_asset_id();
+            coins_to_spend_without_cache(
+                owner,
+                query_per_asset,
+                excluded_ids,
+                max_input,
+                base_asset_id,
+                read_view.as_ref(),
+            )
+            .await
+        }
     }
 }
 
-impl From<CoinModel> for Coin {
-    fn from(value: CoinModel) -> Self {
-        Coin(value)
+async fn coins_to_spend_with_cache(
+    owner: fuel_tx::Address,
+    query_per_asset: Vec<SpendQueryElementInput>,
+    excluded_ids: Option<ExcludeInput>,
+    max_input: u16,
+    db: &ReadView,
+) -> async_graphql::Result<Vec<Vec<CoinType>>> {
+    let mut all_coins = Vec::with_capacity(query_per_asset.len());
+
+    let excluded = ExcludedCoinIds::new(
+        excluded_ids
+            .iter()
+            .flat_map(|exclude| exclude.utxos.iter())
+            .map(|utxo_id| &utxo_id.0),
+        excluded_ids
+            .iter()
+            .flat_map(|exclude| exclude.messages.iter())
+            .map(|nonce| &nonce.0),
+    );
+
+    for asset in query_per_asset {
+        let asset_id = asset.asset_id.0;
+        let total_amount = asset.amount.0;
+        let max = asset
+            .max
+            .and_then(|max| u16::try_from(max.0).ok())
+            .unwrap_or(max_input)
+            .min(max_input);
+
+        let selected_stream = futures::stream::iter(
+            select_coins_to_spend(
+                db.off_chain.coins_to_spend_index(&owner, &asset_id),
+                total_amount,
+                max,
+                &excluded,
+                db.batch_size,
+            )
+            .await?,
+        )
+        .yield_each(db.batch_size);
```

Comment:
The slowest operation is iteration over the storage. Because you already have `Vec<CoinsToSpendIndexEntry>` in the memory, you can just process everything together for one run. It should be very fast.

So we can remove `yield_each` and usage fo the stream here=)